### DISCUSSION TITLE
싱글톤 패턴

### DISCUSSION THREAD

**Discussion by vumrra** [subak-ba/cs-study#13] [디자인 패턴]
# 싱글톤 패턴

싱글톤 패턴은 특정 클래스의 인스턴스를 딱 1개만 생성되는 것을 보장하는 디자인 패턴입니다.

생성자를 통해 여러번 객체의 인스턴스를 가져온다고 하더라고 최초 호출 시 생성된 인스턴스를 반환해주어 메모리에 하나의 인스턴스만 저장되어 관리되는 패턴입니다.

## 싱글톤 패턴의 장점

만약 서비스에 1초에 1만건의 요청이 온다면 해당 요청을 처리하기 위한 똑같은 객체를 1초에 1만번 생성하고 소멸하는 과정에서 메모리 낭비 문제가 발생합니다.

![image](https://github.com/subak-ba/cs-study/assets/131235625/61d410ed-f7e3-4e52-96cb-73c7146320c3)

하지만 싱글톤 패턴을 적용한다면 해당 클래스의 인스턴스는 하나만 생성되도록 보장헤주기 때문에 최초 생성 후 객체를 재사용(공유)하여 메모리 낭비 문제를 방지할 수 있습니다.

![image](https://github.com/subak-ba/cs-study/assets/131235625/d4d4923f-5b34-46ac-a06d-ddcc23d45bee)

## 싱글톤의 구현

아래의 코드는 자바에서 싱글톤을 구현한 예시입니다.

```java
public class Singleton {
    private static final Singleton instance = new Singleton();

    private Singleton() {
    }
    
    public static Singleton getInstance() {
        return instance;
    }
}
```

static영역에 인스턴스 객체를 생성한 이후 인스턴스가 필요할 때 new 키워드로 생성하는 것이 아닌 getInstance() 메서드를 사용하여 최초 생성된 인스턴스를 반환합니다.

- private 접근 제어자를 두어 혹시라도 new 키워드로 새 인스턴스를 생성하는 것을 막습니다.
- getInstance() 메서드가 static 메서드인 이유는 일반 메서드는 인스턴스를 생성한 후에 호출해야하는 반면

```java
// nomal method
Test instance = new Test();
instance.method();
```

- ststic 메서드는 인스턴스 생성없이 호출이 가능하기 때문이다.
    - static 메서드가 해당 클래스의 인스턴스 생성 없이 사용 가능한 이유는 static 키워드가 붙은 필드, 메서드는 메모리의 static 영역에 따로 할당된 후 프로그램이 끝날때 까지 메모리에 값을 유지하기 때문에 공유가 가능하고 인스턴스 생성 없이 접근이 가능합니다.

```java
// static method
Test.method();
```

아래 코드는 자바에서 싱글톤의 간단한 사용 예시입니다.

```java
public static void main(String[] args) {
		Singleton ins1 = Singleton.getInstance();
		Singleton ins2 = Singleton.getInstance();
		
		// ins1 == ins2 -> true
		
}
```

- getInstance() 메서드로 가져온 인스턴스는 몇번을 호출하여 인스턴스를 불러와도 항상 같은 인스턴스를 반환한다.

## 싱글톤 패턴의 문제점

메모리 낭비를 방지해주는 아주 좋은 디자인 패턴인것 처럼 보이는 싱글톤 패턴은 사실 안티 패턴 이라고도 불릴만큼 단점을 많이 가지고 있다.

### 싱글톤 패턴을 구현하는 코드(보일러 플레이트)가 많이 들어간다.

위에서 짧은 예제를 보았다시피 싱글톤을 구현하기 위해서는 많은 코드가 필요로 합니다.

### private 생성자 때문에 자식 클래스를 만들기 어렵다.

- 싱글톤 패턴은 인스턴스가 2개 이상으로 생기는 경우를 막기 위해 기본 생성자를 private 해두기 때문에 상속을 통해 자식 클래스를 만들 수 없습니다. 그렇게 때문에 다형성을 활용할 수 없습니다.
- 상속을 사용하여 자식 클래스를 생성하였을 때 부모 클래스의 기본 생성자가 자동으로 실행되기 때문입니다.

### 높은 의존성

의존 관계상 싱글톤 인스턴스를 필요로하는 클라이언트에서는 싱글톤 클래스를 의존하는 것이 필수적이기 때문에 DIP를 위반합니다.

#2

**Comment by onee-only** [comment]
싱글톤 패턴이 문제점이 있다면 지금 널리 사용되는 방법에는 어떤 것이 있나요?
예를 들어 스프링 프레임워크에서는 어떻게 컴포넌트를 중복되지 않게 관리하나요?
